server/service/service_osquery.go

```diff
@@ -454,22 +452,19 @@ func (svc *Service) detailQueriesForHost(ctx context.Context, host fleet.Host) (
 	return queries, nil
 }
 
-func (svc *Service) shouldUpdate(lastUpdated time.Time, interval time.Duration) bool {
+func (svc *Service) shouldUpdate(lastUpdated time.Time, interval time.Duration, hostID uint) bool {
 	var jitter time.Duration
 	if svc.config.Osquery.MaxJitterPercent > 0 {
-		maxJitter := time.Duration(svc.config.Osquery.MaxJitterPercent) * interval / time.Duration(100.0)
-		randDuration, err := rand.Int(rand.Reader, big.NewInt(int64(maxJitter)))
-		if err == nil {
-			jitter = time.Duration(randDuration.Int64())
-		}
+		maxJitter := int64(svc.config.Osquery.MaxJitterPercent) * int64(interval.Seconds()) / 100.0
+		jitter = time.Duration((int64(hostID) + svc.getJitterSeed()) % maxJitter)
 	}
 	cutoff := svc.clock.Now().Add(-(interval + jitter))
```

Comment:
`jitter` is number of seconds, so this should be `(interval + jitter * time.Second)`, right?

Reply:
Also (no idea if this is the expected behavior) the jitter is always positive, so the interval is always between [interval, interval+maxJitter].

I could be wrong, but was expecting the interval to be between `[interval-maxJitter, interval+maxJitter]`.

Reply:
If we make it only positive, then we make sure to never have overlap.

> jitter is number of seconds, so this should be (interval + jitter * time.Second), right?

Mmmmaybe? I think this might be the way:

```diff
diff --git a/server/service/service_osquery.go b/server/service/service_osquery.go
index 50b5f44f3..1463875eb 100644
--- a/server/service/service_osquery.go
+++ b/server/service/service_osquery.go
@@ -456,7 +456,7 @@ func (svc *Service) shouldUpdate(lastUpdated time.Time, interval time.Duration,
        var jitter time.Duration
        if svc.config.Osquery.MaxJitterPercent > 0 {
                maxJitter := int64(svc.config.Osquery.MaxJitterPercent) * interval.Milliseconds() / 100.0
-               jitter = time.Duration((int64(hostID) + svc.getJitterSeed()) % maxJitter)
+               jitter = time.Duration((int64(hostID)+svc.getJitterSeed())%maxJitter) * time.Millisecond
        }
        cutoff := svc.clock.Now().Add(-(interval + jitter))
        return lastUpdated.Before(cutoff)
```

Reply:
https://go.dev/play/p/aDLE7-jcu1a

```go
	maxJitter := int64(svc.config.Osquery.MaxJitterPercent) * interval / 100.0
	jitter = time.Duration((int64(hostID)+svc.getJitterSeed())%maxJitter)
	// then you can "interval+jitter"
```

Reply:
But interval is in nanoseconds